# Load Dataset

In [114]:
from sklearn.datasets import load_boston

boston = load_boston()
boston

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

# Deskripsi dari Sample Dataset

In [115]:
print(boston.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [116]:
from sklearn import datasets
import pandas as pd

boston_data = datasets.load_boston()
df_boston = pd.DataFrame(boston_data.data,columns=boston_data.feature_names)
df_boston['target'] = pd.Series(boston_data.target)
df_boston.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

# Sample Dataset

Alasan saya mengambil features "PTRATIO" = Jumlah siswa yang bersekolah dibagi dengan jumlah guru di institusi dan features "AGE"= proporsi unit yang ditempati pemilik yang dibangun sebelum tahun 1940. untuk mengetahui perbedaan harga rumah yang jumlah siswa dengan jumlah guru yang proper dan juga proporsi unitnya apakah dibangun sebelum tahun 1940 ataupun sebaliknya.

In [117]:
import pandas as pd

harga_rumah = {'pupil': [15.3, 17.8, 17.8, 18.7, 18.7], 
         'age': [65.2, 78.9, 61.1, 45.8, 54.2],
         'harga': [24, 21.6, 34.7, 33.4, 36.2]}

harga_rumah_df = pd.DataFrame(harga_rumah)
harga_rumah_df

pupil   age  harga
0   15.3  65.2   24.0
1   17.8  78.9   21.6
2   17.8  61.1   34.7
3   18.7  45.8   33.4
4   18.7  54.2   36.2

* pupil dan age berperan sebagai features.
* harga sebagai berperan target.

# Regression With KNN

### Features & Target

In [118]:
import numpy as np

X_train = np.array(harga_rumah_df[['pupil', 'age']]) # sekumpulan nilai features untuk training set
y_train = np.array(harga_rumah_df['harga']) # sekumpulan nilai target untuk training set

print(f'X_train:\n{X_train}\n')
print(f'y_train: {y_train}')

X_train:
[[15.3 65.2]
 [17.8 78.9]
 [17.8 61.1]
 [18.7 45.8]
 [18.7 54.2]]

y_train: [24.  21.6 34.7 33.4 36.2]


### Training KNN Regression Model

In [120]:
from sklearn.neighbors import KNeighborsRegressor

K = 3 
model = KNeighborsRegressor(n_neighbors=K)
model.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=3)

"K" akan berkolerasi dengan jumlah/banyaknya tetangga yang akan digunakan untuk proses prediksi

*Model machine learning yang digunakan untuk regression tasks sering kali dikenal dengan istilah "Regressor" sedangkan model machine learning yang digunakan untuk classification tasks sering kali dikenal dengan istilah "Classifier"*

### Prediksi Harga

In [121]:
X_new = np.array([[19.1, 36]])
X_new

array([[19.1, 36. ]])

* Melakukan prediksi harga berdasarkan pupil dan age

In [122]:
y_pred = model.predict(X_new)
y_pred

array([34.76666667])

* Melakukan prediksi berat badan dengan memanfaatkan model KNN Regressor yang sudah di training sebelumnya
* Data Diprediksi dengan pupil "19.1" dan age "36." yang harganya 34.76666667

### Evaluasi KNN Regression Model

In [140]:
X_test = np.array([[15.3, 65.2], [17.8, 78.9], [17.8, 61.1], [18.7, 45.8],[18.7, 54.2]])
y_test = np.array([25, 22.3, 35.1, 34.2, 37.6])

print(f'X_test:\n{X_test}\n')
print(f'y_test: {y_test}')

X_test:
[[15.3 65.2]
 [17.8 78.9]
 [17.8 61.1]
 [18.7 45.8]
 [18.7 54.2]]

y_test: [25.  22.3 35.1 34.2 37.6]


Baris Pertama : Menyiapkan 4 instances atau data point

In [141]:
y_pred = model.predict(X_test)
y_pred

array([31.63333333, 26.76666667, 31.63333333, 34.76666667, 34.76666667])

Hasilnya bahwa data pupil "15.3" dan age "65.2",ini diprediksi memilki harga "31.63333333" sedangkan data yang yang diharapkan adalah "25"

#### Coefficient of Determination atau $R^2$ 

In [142]:
from sklearn.metrics import r2_score

r_squared = r2_score(y_test, y_pred)

print(f'R-squared: {r_squared}')

R-squared: 0.5371468233927887


R-squared semakin dia mendekati 1 maka semakin baik dan semakin dia mendekati 0 atau bahkan ketika nilainya negatif ini mengindikasikan model yang kurang baik

#### Mean Absolute Error (MAE) atau Mean Absolute Deviation (MAD)

In [143]:
from sklearn.metrics import mean_absolute_error

MAE = mean_absolute_error(y_test, y_pred)

print(f'MAE: {MAE}')

MAE: 3.5933333333333324


*Semakin kecil nilai MAE akan mengindikasikan kualitas model yang makin/lebih baik*

#### Mean Squared Error (MSE) atau Mean Squared Deviation (MSD)

In [144]:
from sklearn.metrics import mean_squared_error

MSE = mean_squared_error(y_test, y_pred)

print(f'MSE: {MSE}')

MSE: 16.863777777777777


Semakin kecil nilai MSE akan mengindikasikan kualitas model yang makin/lebih baik

*Nilai MSE akan selalu lebih besar bila dibandingkan dengan nilai MAE karena setiap nilai errornya untuk MSE akan dipangkatkan dua*

### Menerapkan Features Scaling pada KNN

#### Dataset

In [155]:
import numpy as np
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

# Training Set
X_train = np.array([[15, 65.2], [17., 78.9], [17., 61.1], [18., 45.8],[18., 54.2]])
y_train = np.array([26.7, 24.4, 34.2, 35.6, 36.2])


# Test Set
X_test = np.array([[15.3, 65.2], [17.8, 78.9], [17.8, 61.1], [18.7, 45.8],[18.7, 54.2]])
y_test = np.array([25, 22.3, 35.1, 34.2, 37.6])

Training dan Testing set yang kita gunakan ini sama persis dengan Training dan Testing set yang kita gunakan sebelumnya

#### Features Scaling (Standard Scaler)

Training dan Testing setnya tersedia kita tidak langsung melakukan proses training model.tetapi pertama-tama kita akan scalling dulu featuresnya

In [156]:
X_train_scaled = ss.fit_transform(X_train)
X_test_scaled = ss.transform(X_test)

print(f'X_train_scaled:\n{X_train_scaled}\n')
print(f'X_test_scaled:\n{X_test_scaled}\n')

X_train_scaled:
[[-1.82574186  0.37499336]
 [ 0.          1.60994747]
 [ 0.          0.00540856]
 [ 0.91287093 -1.37377377]
 [ 0.91287093 -0.61657563]]

X_test_scaled:
[[-1.55188058  0.37499336]
 [ 0.73029674  1.60994747]
 [ 0.73029674  0.00540856]
 [ 1.55188058 -1.37377377]
 [ 1.55188058 -0.61657563]]



#### Training & Evaluasi Model

In [157]:
model.fit(X_train_scaled, y_train) #features yang sudah kita scalling
y_pred = model.predict(X_test_scaled)

MAE = mean_absolute_error(y_test, y_pred)
MSE = mean_squared_error(y_test, y_pred)

print(f'MAE: {MAE}')
print(f'MSE: {MSE}')

MAE: 3.2733333333333348
MSE: 20.950888888888908


Perbandingan Hasil MAE dan MSE setelah menerapkan features scaling hasilnya lebih kecil atau dengan kata lain kita bisa menghasilkan model dengan kualitas atau performa yang lebih setelah menerapkan features scaling